core/store/src/trie/iterator.rs

```diff
@@ -367,15 +375,24 @@ impl<'a> Iterator for TrieIterator<'a> {
     fn next(&mut self) -> Option<Self::Item> {
         loop {
             let iter_step = self.iter_step()?;
+
+            let can_process = match self.max_depth {
+                Some(max_depth) => self.key_nibbles.len() <= max_depth,
+                None => true,
+            };
+            if !can_process {
+                continue;
+            }
+
```

Comment:
Note for reviewers:
There are many places where this if statement could go but here at least it's only in a single place and handles all the node types. There is minor perf penalty for doing it here - here we already have the nodes exceeding max depth lined up for iteration. I feel this is much better becuase it's cleaner and it's not perf-critical.